diff --git a/src/manager/event_handlers.ts b/src/manager/event_handlers.ts
index 139f78d..89e843e 100644
--- a/src/manager/event_handlers.ts
+++ b/src/manager/event_handlers.ts
@@ -27,25 +27,29 @@ import {
     getRoundedCornersCfg,
     getRoundedCornersEffect,
     shouldEnableEffect,
-    unwrapActor,
     updateShadowActorStyle,
     windowScaleFactor,
 } from './utils.js';
 
 export function onAddEffect(actor: RoundedWindowActor) {
-    logDebug(`Adding effect to ${actor?.metaWindow.title}`);
+    const win = actor?.metaWindow;
+    if (!win) {
+        return;
+    }
 
-    const win = actor.metaWindow;
+    logDebug(`Adding effect to ${win.title}`);
 
     if (!shouldEnableEffect(win)) {
         logDebug(`Skipping ${win.title}`);
         return;
     }
 
-    unwrapActor(actor)?.add_effect_with_name(
-        ROUNDED_CORNERS_EFFECT,
-        new RoundedCornersEffect(),
-    );
+    actor
+        .get_last_child()
+        ?.add_effect_with_name(
+            ROUNDED_CORNERS_EFFECT,
+            new RoundedCornersEffect(),
+        );
 
     const shadow = createShadow(actor);
 
@@ -78,10 +82,10 @@ export function onAddEffect(actor: RoundedWindowActor) {
 
 export function onRemoveEffect(actor: RoundedWindowActor): void {
     const name = ROUNDED_CORNERS_EFFECT;
-    unwrapActor(actor)?.remove_effect_by_name(name);
+    actor?.get_last_child()?.remove_effect_by_name(name);
 
     // Remove shadow actor
-    const shadow = actor.rwcCustomData?.shadow;
+    const shadow = actor?.rwcCustomData?.shadow;
     if (shadow) {
         global.windowGroup.remove_child(shadow);
         shadow.clear_effects();
@@ -89,19 +93,21 @@ export function onRemoveEffect(actor: RoundedWindowActor): void {
     }
 
     // Remove all timeout handler
-    const timeoutId = actor.rwcCustomData?.unminimizedTimeoutId;
+    const timeoutId = actor?.rwcCustomData?.unminimizedTimeoutId;
     if (timeoutId) {
         GLib.source_remove(timeoutId);
     }
-    delete actor.rwcCustomData;
+    if (actor) {
+        delete actor.rwcCustomData;
+    }
 }
 
 export function onMinimize(actor: RoundedWindowActor): void {
     // Compatibility with "Compiz alike magic lamp effect".
     // When minimizing a window, disable the shadow to make the magic lamp effect
     // work.
-    const magicLampEffect = actor.get_effect('minimize-magic-lamp-effect');
-    const shadow = actor.rwcCustomData?.shadow;
+    const magicLampEffect = actor?.get_effect('minimize-magic-lamp-effect');
+    const shadow = actor?.rwcCustomData?.shadow;
     const roundedCornersEffect = getRoundedCornersEffect(actor);
     if (magicLampEffect && shadow && roundedCornersEffect) {
         logDebug('Minimizing with magic lamp effect');
@@ -114,8 +120,8 @@ export function onUnminimize(actor: RoundedWindowActor): void {
     // Compatibility with "Compiz alike magic lamp effect".
     // When unminimizing a window, wait until the effect is completed before
     // showing the shadow.
-    const magicLampEffect = actor.get_effect('unminimize-magic-lamp-effect');
-    const shadow = actor.rwcCustomData?.shadow;
+    const magicLampEffect = actor?.get_effect('unminimize-magic-lamp-effect');
+    const shadow = actor?.rwcCustomData?.shadow;
     const roundedCornersEffect = getRoundedCornersEffect(actor);
     if (magicLampEffect && shadow && roundedCornersEffect) {
         shadow.visible = false;
@@ -197,9 +203,9 @@ function createShadow(actor: Meta.WindowActor): St.Bin {
  * @param actor - The window actor to refresh the shadow for.
  */
 function refreshShadow(actor: RoundedWindowActor) {
-    const win = actor.metaWindow;
-    const shadow = actor.rwcCustomData?.shadow;
-    if (!shadow) {
+    const win = actor?.metaWindow;
+    const shadow = actor?.rwcCustomData?.shadow;
+    if (!win || !shadow) {
         return;
     }
 
@@ -218,7 +224,10 @@ function refreshShadow(actor: RoundedWindowActor) {
  * @param actor - The window actor to refresh the rounded corners settings for.
  */
 function refreshRoundedCorners(actor: RoundedWindowActor): void {
-    const win = actor.metaWindow;
+    const win = actor?.metaWindow;
+    if (!win) {
+        return;
+    }
 
     const windowInfo = actor.rwcCustomData;
     const effect = getRoundedCornersEffect(actor);
diff --git a/src/manager/event_manager.ts b/src/manager/event_manager.ts
index 43c3632..72b4233 100644
--- a/src/manager/event_manager.ts
+++ b/src/manager/event_manager.ts
@@ -130,10 +130,20 @@ function applyEffectTo(actor: RoundedWindowActor) {
     // In wayland sessions, the surface actor of XWayland clients is sometimes
     // not ready when the window is created. In this case, we wait until it is
     // ready before applying the effect.
-    if (!actor.firstChild) {
-        const id = actor.connect('notify::first-child', () => {
-            applyEffectTo(actor);
-            actor.disconnect(id);
+    // For Blur My Shell compatibility, we have to make sure that the actor is
+    // not a bms-application-blurred-widget. This actor is created by BMS and
+    // is not the actual window actor.
+    const bmsActorName = 'bms-application-blurred-widget';
+    const actorReady =
+        (actor.firstChild && actor.firstChild.name !== bmsActorName) ||
+        (actor.lastChild && actor.lastChild.name !== bmsActorName);
+
+    if (!actorReady) {
+        const id = actor.connect('child-added', (actor, child) => {
+            if (child.name !== bmsActorName) {
+                applyEffectTo(actor);
+                actor.disconnect(id);
+            }
         });
 
         return;
@@ -193,7 +203,9 @@ function applyEffectTo(actor: RoundedWindowActor) {
  */
 function removeEffectFrom(actor: RoundedWindowActor) {
     disconnectAll(actor);
-    disconnectAll(actor.metaWindow);
+    if (actor?.metaWindow) {
+        disconnectAll(actor.metaWindow);
+    }
 
     handlers.onRemoveEffect(actor);
 }
diff --git a/src/manager/utils.ts b/src/manager/utils.ts
index 0b26b24..55e009c 100644
--- a/src/manager/utils.ts
+++ b/src/manager/utils.ts
@@ -18,19 +18,6 @@ import {readFile} from '../utils/file.js';
 import {logDebug} from '../utils/log.js';
 import {getPref} from '../utils/settings.js';
 
-/**
- * Get the actor that rounded corners should be applied to.
- * In Wayland, the effect is applied to WindowActor, but in X11, it is applied
- * to WindowActor.first_child.
- *
- * @param actor - The window actor to unwrap.
- * @returns The correct actor that the effect should be applied to.
- */
-export function unwrapActor(actor: Meta.WindowActor): Clutter.Actor | null {
-    const type = actor.metaWindow.get_client_type();
-    return type === Meta.WindowClientType.X11 ? actor.get_first_child() : actor;
-}
-
 /**
  * Get the correct rounded corner setting for a window (custom settings if a
  * window has custom overrides, global settings otherwise).
@@ -67,11 +54,8 @@ type RoundedCornersEffectType = InstanceType<typeof RoundedCornersEffect>;
 export function getRoundedCornersEffect(
     actor: Meta.WindowActor,
 ): RoundedCornersEffectType | null {
-    const win = actor.metaWindow;
     const name = ROUNDED_CORNERS_EFFECT;
-    return win.get_client_type() === Meta.WindowClientType.X11
-        ? (actor.firstChild.get_effect(name) as RoundedCornersEffectType)
-        : (actor.get_effect(name) as RoundedCornersEffectType);
+    return actor?.lastChild?.get_effect(name) as RoundedCornersEffectType | null;
 }
 
 /**
@@ -116,8 +100,8 @@ export function computeBounds(
     // clip its shadow and recompute the outer bounds for it.
     if (
         getPref('tweak-kitty-terminal') &&
-        actor.metaWindow.get_client_type() === Meta.WindowClientType.WAYLAND &&
-        actor.metaWindow.get_wm_class_instance() === 'kitty'
+        actor.metaWindow?.get_client_type() === Meta.WindowClientType.WAYLAND &&
+        actor.metaWindow?.get_wm_class_instance() === 'kitty'
     ) {
         const [x1, y1, x2, y2] = APP_SHADOWS.kitty;
         const scale = windowScaleFactor(actor.metaWindow);
